python/ray/autoscaler/v2/scheduler.py

```diff
@@ -417,15 +516,223 @@ def schedule(self, request: SchedulingRequest) -> SchedulingReply:
+        nodes: List[SchedulingNode],
+        num_to_terminate: int,
+        cause: TerminationRequest.Cause,
+        max_num_nodes: Optional[int] = None,
+        max_num_nodes_per_type: Optional[int] = None,
+        idle_duration_ms: Optional[int] = None,
+    ) -> Tuple[List[SchedulingNode], List[SchedulingNode]]:
+        """
+        Terminate the nodes.
+
+        Args:
+            nodes: The nodes to be terminated.
+            num_to_terminate: The number of nodes to be terminated.
+            cause: The cause of the termination.
+
+            max_num_nodes: The max number of nodes for the entire cluster only
+                used when the cause is TerminationRequest.Cause.MAX_NUM_NODES.
+            max_num_nodes_per_type: The max number of nodes for each node type.
+                Only used when the cause is
+                TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE.
+            idle_duration_ms: The idle duration in ms. Only used when the cause is
+                TerminationRequest.Cause.IDLE.
+
+        Returns:
+            A tuple of:
+                - The terminated nodes.
+                - The remained nodes.
+        """
+
+        terminated_nodes, remained_nodes = (
+            nodes[:num_to_terminate],
+            nodes[num_to_terminate:],
+        )
+
+        for node in terminated_nodes:
+            node.status = SchedulingNodeStatus.TO_TERMINATE
+            node.termination_request = TerminationRequest(
+                id=str(time.time_ns()),
+                instance_id=node.im_instance_id,
+                ray_node_id=node.ray_node_id,
+                cause=cause,
+            )
+            if cause == TerminationRequest.Cause.MAX_NUM_NODES:
+                node.termination_request.max_num_nodes = max_num_nodes
+            elif cause == TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE:
+                node.termination_request.max_num_nodes_per_type = max_num_nodes_per_type
+            elif cause == TerminationRequest.Cause.IDLE:
+                node.termination_request.idle_duration_ms = idle_duration_ms
+            else:
+                raise ValueError("Unknown termination cause: {}".format(cause))
+
+        return terminated_nodes, remained_nodes
+
+    @staticmethod
+    def _sort_nodes_for_termination(node: SchedulingNode) -> Tuple:
+        """
+        Sort the nodes for termination increasingly by:
+
+            1. First if it doesn't have a ray node.
+            2. Then if the nodes are idle
+            3. Then with lower resources util nodes first.
+
+        Such that nodes sorted earlier will be terminated first.
+        """
+
+        running_ray = node.status == SchedulingNodeStatus.RUNNING
+        # Reverse the idle duration such that the nodes with the largest idle duration
+        # will be terminated first.
+        idle_dur = -1 * node.idle_duration_ms
+
+        utils_per_resources = {}
+        for resource, total in node.total_resources.items():
```

Comment:
is this something new?
i thought we don't terminate nodes if there are utilizations to avoid application failures 
cc @jjyao @rkooo567 to confirm

Reply:
this is not something new - we have been killing nodes in v1 for nodes that are:
1. with outdated configs (detected from config hash changes)
2. would exceed max_num_workers for the node type 
3. would exceed max_workers for all nodes.